File: algo1/2/q.php
<?php

new QuickSort();

class QuickSort {
    public $arr = array();
    //public $comp = 0;

    function __construct() {
        echo '<pre>';

        $a = file('IntegerArray.txt');
        function toInt($x) { return (int)$x; }

        $this->arr = array_map( 'toInt', $a );

        $c = count( $this->arr );
        $this->quicksort( 0, $c - 1 );

        $bound = ceil( $c * log( $c, 2 ) );
        echo "Upper bound: <b>{$bound}</b><br/>";
        echo "Comparisons: <b>{$this->comp}</b><br /><br />";

        echo '</pre>';
    }

    function quicksort( $l, $r ) {
        // Accumulate comparisons
        $this->comp += ( $r - $l );

        // Chose pivot
        $p = $this->choosepivot( $l, $r );
        $pivot = $this->arr[$p];

        // Move pivot on first position
        $this->swap( $l, $p );

        // Partition
        $i = $l + 1;
        for( $j = $i; $j <= $r; $j++ ) {
            if( $this->arr[$j] < $pivot ) {
                $this->swap( $i, $j );
                $i++;
            }
        }

        // Put pivot back in its rightful position
        $this->swap( $l, $i-1 );

        // Recurse on 1st half
        if( $i - 2 > $l ) $this->quicksort( $l, $i-2 );

        // Recurse on 2nd half
        if( $r > $i ) $this->quicksort( $i, $r );
    }

    function choosepivot( $l, $r ) {
        if( $r - $l == 1 ) return $l;

        // Pivot is first element
        //return $l;

        // Pivot is last element
        //return $r;

        // Flip coins
        //return rand( $l, $r );

        // Median rule

        //$m = floor( ($l + $r) / 2 );
        $m = rand( $l+1, $r-1 );

        $left = $this->arr[$l];
        $right = $this->arr[$r];
        $center = $this->arr[$m];

        $a = $left;
        $b = $right;
        $c = $center;

        // if( $left > $center ) $this->swap( $l, $m );
        // if( $left > $right ) $this->swap( $l, $r );
        // if( $center > $right ) $this->swap( $m, $r );

        // return $m;

        if( ( $c < $a and $a < $b ) or ( $b < $a and $a < $c ) ) return $l;
        if( ( $a < $b and $b < $c ) or ( $c < $b and $b < $a ) ) return $r;
        if( ( $b < $c and $c < $a ) or ( $a < $c and $c < $b ) ) return $m;
    }

    function swap( $x, $y ) {
        $t = $this->arr[$x];
        $this->arr[$x] = $this->arr[$y];
        $this->arr[$y] = $t;
    }

}

?>

File: _coursera/algo1/2/q.py
import sys
import os
import time

from random import randrange
from random import randint
from math import floor


def qsort1(list):
    if list == []:
        return []

    global comp
    l = len(list)
    comp += ( l - 1 )

    # Choose pivot
    p = choosepivot(list,l)
    pivot = list[p]

    # Put it on 0th position
    swap(list,0,p)

    # Partition
    i = 1
    for j in range(1,l):
        if list[j] < pivot:
            swap(list,i,j)
            i+=1

    # Put pivot back
    swap(list,0,i-1)

    # Recurse on first half
    lesser = qsort1(list[:i-1])

    # Recurse on 2nd half
    greater = qsort1(list[i:])

    return lesser + greater


def qsort2(list):
    if list == []:
        return []

    global comp
    l = len(list)
    comp += ( l - 1 )

    pivot = list.pop(choosepivot(list,l))

    lesser = qsort2([l for l in list if l < pivot])
    greater = qsort2([l for l in list if l >= pivot])
    return lesser + [pivot] + greater

def choosepivot(list,lng):
    # Random
    # return randrange(lng)

    # Left
    # return 0

    # Right
    # return lng-1


    if lng <= 3:
        return 0
    # Median of three (random)
    m = randrange(1, lng-2)

    # Median of three (middle)
    # m = int(floor((lng-1)/2))

    c = list[m]

    [s,l] = [0,lng-1] if list[0] < list[lng-1] else [lng-1,0]

    if c > list[l]:
        return l
    elif c < list[s]:
        return s
    else:
        return m

def swap(arr,x,y):
    t,arr[x] = arr[x],arr[y]
    arr[y] = t



t0 = time.clock()

comp = 0
arr = [int(line) for line in open('IntegerArray.txt')]
qsort1(arr)
print 'comparisons: ' + str(comp)
print 'time: ' + str(time.clock()-t0)



t0 = time.clock()

comp = 0
arr = [int(line) for line in open('IntegerArray.txt')]
qsort2(arr)
print 'comparisons: ' + str(comp)
print 'time: ' + str(time.clock()-t0)

t0 = time.clock()

arr = [int(line) for line in open('IntegerArray.txt')]
arr.sort()
print 'time: ' + str(time.clock()-t0)